### PR TITLE
Preallocate journal file space. Avoid writing metadata on every sync.

### PR DESCRIPTION
**Pull request by shvachko**
Hey guys.
I am new here, but WANdisco was using Prevayler in production for quite a while.
I thought you may find my patch interesting, as it increases Prevayler's performs 2x.

### CODE REVIEW

**Comment by shvachko** [comment]
Here is more details:

I was comparing Prevayler's performance with HDFS NameNode EditsLog. The two systems serve the same purpose, that is journalling transactions. And my assumption was that besides all the differences in implementation details, if I put enough load on either of them to make them i/o bound, the performance should be roughly the same. Unless of course something is going wrong with one of the systems.

In a nutshell, my benchmarks showed that Prevayler's throughput (tx/sec) is two times lower than that of a NameNode. After some digging I realized that the main reason for that is that NameNode does not write metadata on every sync, only the data. This is done by using
`FileChannel.force(false)`
instead of
`FileDescriptor.sync()`
The latter is used in `DurableOutputStream.waitUntilSynced()`, and it syncs both data and metadata, causing two IOs instead of one as in `force(false)`.
In general one wants to save file metadata when new data is appended to a file. But if the disk space for the journal file is preallocated in advance then there is no need to save metadata frequently, as file size is not changing. This is another difference that NameNode's EditsLog has compared to Prevayler.

For benchmarking I used `NNThroughputBenchmark`, which is a standard tool in HDFS, and which measures the throughput of different NameNode operations excluding any network overhead (no RPCs). I wrote a similar `PrevaylerBenchmark`, which uses a map as a prevalent system, and adds key-value pairs into it as transactions. The benchmark allows to run multiple update threads. Here are some results comparing Prevayler with NameNode before and after my changes:

o | Threads | #Ops | Throughput Ops/sec | Elapsed time sec
--- | ---------- | ------ | ------------------------- | ----------------------
Prevayler | 1 | 5000 | 40.486 | 123.498
NameNode | 1 | 5000 | 77.438 | 64.568
Prevayler | 512 | 1,000,000 | 9439.929 | 105.933
NameNode | 512 | 1,000,000 | 13306.0116 | 64.568

After the patch:

o | Threads | #Ops | Throughput Ops/sec | Elapsed time sec
--- | ---------- | ------ | ------------------------- | ----------------------
Prevayler | 1 | 5000 | **81.904** | 61.047
NameNode | 1 | 5000 | 77.438 | 64.568
Prevayler | 512 | 1,000,000 | **16230.604** | 61.612
NameNode | 512 | 1,000,000 | 13306.0116 | 64.568

The patch contains `PrevaylerBenchmark` and changes to syncing and preallocation.
Please take a look.

File: core/src/main/java/org/prevayler/foundation/DurableOutputStream.java [hunk 1]
```diff
@@ -68,13 +74,36 @@
...
+    if (need <= 0) {
+      return;
+    }
+    while (need > 0) {
+      ByteBuffer buf = ByteBuffer.allocateDirect(_preallocateLength);
+      int written = _fileChannel.write(buf, size);
+      assert written == _preallocateLength : "incomplete write";
+      need -= written;
+    }
+    _fileDescriptor.sync();
```
Comment (klauswuestefeld) [inline core/src/main/java/org/prevayler/foundation/DurableOutputStream.java]: Why is sync() used here and force(false) used below?

**Comment by shvachko** [comment]
`preallocate()` changes file size, so it should sync the new size to disk by calling `_fileDescriptor.sync();`. Alternatively one could call `_fileChannel.force(true)` with the same effect.
Once the file size is set there is no need to update metadata (until the next `preallocate()`), therefore `sync()` below calls `_fileChannel.force(false)`, which sync only the data and avoids the second i/o.

**Comment by klauswuestefeld** [comment]
Could you please change the sync() to _fileChannel.force(true) ? It will be more symmetrical that way. I can then merge it. I like this optimization very much, Konstantin. :)

**Comment by shvachko** [comment]
I just did. Also removed _fileDescriptor as it is not used anywhere any more.
Thanks, Klaus.

**Comment by felipecruz** [comment]
Hello @shvachko 

Nice patch! Very impressive!

Little "details" can have a big impact on overall performance. I would love to hear more about how WANDisco uses Prevayler in production if it's possible :)

**Comment by sormuras** [comment]
Like the patch - and will update the comparsion to P8 ... later this year.

http://codeturm.blogspot.de/2014/11/p8-alternative-prevayler-implementation.html

**Pull request merged**